eogrow/pipelines/download.py

```diff
@@ -135,6 +170,20 @@ def get_execution_arguments(self, workflow: EOWorkflow) -> List[Dict[EONode, Dic
 
         return exec_args
 
+    def run_procedure(self) -> Tuple[List[str], List[str]]:
+        execution_kind = self._prepare_execution()
+        session_loader = self._create_session_loader(execution_kind)
```

Comment:
Because calling `self._prepare_execution` doesn't really cost anything, you could restructure this a bit so that `self._create_session_loader` doesn't have to receive `execution_type` but instead just calls `self._prepare_execution` itself. This means that `self._create_session_loader` could be called only in `_get_download_node` method and not sooner. 

- advantage: you wouldn't have to propagate `session_loader` parameters 2 levels deep,
- disadvantage: `self._prepare_execution` gets called 1 more time.

Reply:
because `_prepare_execution` has side-effects I went with value propagation. Another option is to assign the result to an attribute and read from it in `_get_download_node` but then the execution model becomes more complex since the two methods communicate over non-standard channels

Reply:
Yes, but the side effect can happen multiple times without problems. Ray starts a connection with cluster only the first time.

Although I see also your point that we might want to be more explicit with this and not to hide the side-effect in some sub-method. So I'm leaving it up to you to decide.

Reply:
I was thinking that perhaps each pipeline has an attribute `execution_mode` (by default 'single'), and that is *adjusted* in the `_prepare_execution`. However this has precisely the issue described above, if the adjustment hasn't happened yet the value is wrong. The big issue here is that with 'auto' we don't know if ray is available or not. The price of comfort :)

I would leave it as it is, because it also explicitly states the requirement that the workflow is adjusted according to the execution mode, perhaps a good thing.

Reply:
Ok, let's leave it as it is.